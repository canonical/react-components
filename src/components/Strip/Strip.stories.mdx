import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import Col from "../Col";
import Row from "../Row";
import Strip from "./Strip";

<Meta
  title="Strip"
  component={Strip}
  argTypes={{
    children: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => <Strip {...args} />;

### Strip

This is a [React](https://reactjs.org/) component for the Vanilla [Strip](https://docs.vanillaframework.io/patterns/strip/).

The strip pattern provides a full width strip container in which to wrap a row.

### Props

<ArgsTable of={Strip} />

### Light strip

<Canvas>
  <Story
    name="Light strip"
    args={{
      type: "light",
      children: <p>This is a light row</p>,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Dark strip

<Canvas>
  <Story name="Dark strip">
    <Strip type="dark">
      <p>This is a dark row</p>
    </Strip>
  </Story>
</Canvas>

### Accent strip

<Canvas>
  <Story name="Accent strip">
    <Strip type="accent" element="section" includeCol={false}>
      <Col size={8}>
        <h1>Still running Ubuntu 14.04 LTS?</h1>
        <p>
          Learn how to maintain ongoing security compliance for your Ubuntu
          14.04 LTS systems.
        </p>
      </Col>
      <Col size={4}>
        <img
          src="https://assets.ubuntu.com/v1/2217d1c8-Security.svg"
          alt="Placeholder"
        />
      </Col>
    </Strip>
  </Story>
</Canvas>

### Image strip

<Canvas>
  <Story name="Image strip">
    {() => (
      <>
        <Strip
          background="https://assets.ubuntu.com/sites/ubuntu/latest/u/img/backgrounds/image-background-paper.png"
          includeCol={false}
          element="section"
          light
          rowClassName="u-vertically-center"
          type="image"
        >
          <Col size={8}>
            <h1>Get started with big software, fast</h1>
            <p>
              conjure-up lets you summon up a big-software stack as a “spell” —
              a model of the stack, combined with extra know-how to get you from
              an installed stack to a fully usable one. Start using your big
              software instead of learning how to deploy it.
            </p>
          </Col>
          <Col size={4} className="u-hide--small u-align--center">
            <img
              src="https://assets.ubuntu.com/v1/1abb8716-conjure-up-illustration.svg"
              alt="Placeholder"
            />
          </Col>
        </Strip>
        <Strip
          background="https://assets.ubuntu.com/v1/9b68976e-Aubergine_suru_background_2.png"
          dark
          includeCol={false}
          element="section"
          type="image"
        >
          <Row className="u-vertically-center">
            <Col size={8}>
              <h1>We are Canonical</h1>
              <p>
                It is our mission to make open source software available to
                people everywhere. We believe the best way to fuel innovation is
                to give the innovators the technology they need.
              </p>
            </Col>
            <Col size={4} className="u-hide--small u-align--center">
              <img
                src="https://assets.ubuntu.com/v1/9c74eb2d-logo-canonical-white.svg"
                alt="Placeholder"
              />
            </Col>
          </Row>
        </Strip>
      </>
    )}
  </Story>
</Canvas>

### Bordered strip

<Canvas>
  <Story name="Bordered strip">
    <Strip bordered colSize={8} element="section">
      <h2>The node lifecycle</h2>
      <p>
        Each machine (“node”) managed by MAAS goes through a lifecycle — from
        its enlistment or onboarding to MAAS, through commissioning when we
        inventory and can setup firmware or other hardware-specific elements,
        then allocation to a user and deployment, and finally they are released
        back to the pool or retired altogether.
      </p>
    </Strip>
  </Story>
</Canvas>

### Deep strip

<Canvas>
  <Story name="Deep strip">
    <Strip
      deep
      includeCol={false}
      element="section"
      type="light"
      rowClassName="u-vertically-center"
    >
      <Col size={8}>
        <h2>The fastest way to go from development to production in IoT</h2>
        <p>
          Learn about how Ubuntu Core and snaps can help you build your
          connected devices.
        </p>
      </Col>
      <Col size={4} className="u-hide--small u-align--center">
        <img
          src="https://assets.ubuntu.com/v1/808a4e5b-iot.png?h=300"
          alt="Placeholder"
        />
      </Col>
    </Strip>
  </Story>
</Canvas>

### Shallow strip

<Canvas>
  <Story name="Shallow strip">
    <Strip
      shallow
      includeCol={false}
      element="section"
      type="light"
      rowClassName="u-vertically-center"
    >
      <Col size={8}>
        <h2>The fastest way to go from development to production in IoT</h2>
        <p>
          Learn about how Ubuntu Core and snaps can help you build your
          connected devices.
        </p>
      </Col>
      <Col size={4} className="u-hide--small u-align--center">
        <img
          src="https://assets.ubuntu.com/v1/808a4e5b-iot.png?h=300"
          alt="Placeholder"
        />
      </Col>
    </Strip>
  </Story>
</Canvas>
