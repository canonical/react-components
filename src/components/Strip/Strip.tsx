import classNames from "classnames";
import React from "react";
import type { ElementType, HTMLProps, ReactNode } from "react";

import Col from "../Col";
import type { ColSize } from "../Col/Col";
import Row from "../Row";

import type { ClassName, PropsWithSpread } from "types";

export type Props = PropsWithSpread<
  {
    /**
     * The content of the strip.
     */
    children: ReactNode;
    /**
     * A background images for the strip.
     */
    background?: string;
    /**
     * Whether the strip should display borders.
     */
    bordered?: boolean;
    /**
     * Optional classes for the strip.
     */
    className?: ClassName;
    /**
     * The width of the column if `includeCol` has been set.
     */
    colSize?: ColSize;
    /**
     * Whether the strip should be dark.
     */
    dark?: boolean;
    /**
     * Whether the strip should be deep.
     */
    deep?: boolean;
    /**
     * The base HTML element of the strip.
     */
    element?: ElementType;
    /**
     * Whether the strip should wrap the content in a column.
     */
    includeCol?: boolean;
    /**
     * Whether the strip should be light.
     */
    light?: boolean;
    /**
     * Optional classes to apply to the row.
     */
    rowClassName?: string;
    /**
     * Whether the strip should be shallow.
     */
    shallow?: boolean;
    /**
     * The type of the strip (e.g. "accent" or "image").
     */
    type?: string;
  },
  HTMLProps<HTMLElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Strip](https://docs.vanillaframework.io/patterns/strip/).
 *
 * The strip pattern provides a full width strip container in which to wrap a row.
 */
const Strip = ({
  background,
  bordered = false,
  children,
  className,
  colSize = 12,
  dark = false,
  deep = false,
  includeCol = true,
  element: Component = "div",
  light = false,
  rowClassName,
  shallow = false,
  type,
  ...props
}: Props): JSX.Element => (
  <Component
    className={classNames(className, {
      [`p-strip--${type}`]: !!type,
      "p-strip": !type,
      "is-bordered": bordered,
      "is-dark": dark,
      "is-deep": deep,
      "is-light": light,
      "is-shallow": shallow,
    })}
    style={
      background && {
        backgroundImage: `url('${background}')`,
      }
    }
    {...props}
  >
    <Row className={rowClassName}>
      {includeCol ? <Col size={colSize}>{children}</Col> : children}
    </Row>
  </Component>
);

export default Strip;
