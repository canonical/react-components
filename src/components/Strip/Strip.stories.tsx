import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Col from "../Col";
import Row from "../Row";
import Strip from "./Strip";

const meta: Meta<typeof Strip> = {
  component: Strip,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Strip>;

export const LightStrip: Story = {
  name: "Light strip",

  args: {
    type: "light",
    children: <p>This is a light row</p>,
  },
};

export const DarkStrip: Story = {
  render: () => (
    <Strip type="dark">
      <p>This is a dark row</p>
    </Strip>
  ),

  name: "Dark strip",
};

export const AccentStrip: Story = {
  render: () => (
    <Strip type="accent" element="section" includeCol={false}>
      <Col size={8}>
        <h1>Still running Ubuntu 14.04 LTS?</h1>
        <p>
          Learn how to maintain ongoing security compliance for your Ubuntu
          14.04 LTS systems.
        </p>
      </Col>
      <Col size={4}>
        <img
          src="https://assets.ubuntu.com/v1/2217d1c8-Security.svg"
          alt="Placeholder"
        />
      </Col>
    </Strip>
  ),

  name: "Accent strip",
};

export const ImageStrip: Story = {
  render: () => (
    <>
      <Strip
        background="https://assets.ubuntu.com/sites/ubuntu/latest/u/img/backgrounds/image-background-paper.png"
        includeCol={false}
        element="section"
        light
        rowClassName="u-vertically-center"
        type="image"
      >
        <Col size={8}>
          <h1>Get started with big software, fast</h1>
          <p>
            conjure-up lets you summon up a big-software stack as a “spell” — a
            model of the stack, combined with extra know-how to get you from an
            installed stack to a fully usable one. Start using your big software
            instead of learning how to deploy it.
          </p>
        </Col>
        <Col size={4} className="u-hide--small u-align--center">
          <img
            src="https://assets.ubuntu.com/v1/1abb8716-conjure-up-illustration.svg"
            alt="Placeholder"
          />
        </Col>
      </Strip>
      <Strip
        background="https://assets.ubuntu.com/v1/9b68976e-Aubergine_suru_background_2.png"
        dark
        includeCol={false}
        element="section"
        type="image"
      >
        <Row className="u-vertically-center">
          <Col size={8}>
            <h1>We are Canonical</h1>
            <p>
              It is our mission to make open source software available to people
              everywhere. We believe the best way to fuel innovation is to give
              the innovators the technology they need.
            </p>
          </Col>
          <Col size={4} className="u-hide--small u-align--center">
            <img
              src="https://assets.ubuntu.com/v1/9c74eb2d-logo-canonical-white.svg"
              alt="Placeholder"
            />
          </Col>
        </Row>
      </Strip>
    </>
  ),

  name: "Image strip",
};

export const BorderedStrip: Story = {
  render: () => (
    <Strip bordered colSize={8} element="section">
      <h2>The node lifecycle</h2>
      <p>
        Each machine (“node”) managed by MAAS goes through a lifecycle — from
        its enlistment or onboarding to MAAS, through commissioning when we
        inventory and can setup firmware or other hardware-specific elements,
        then allocation to a user and deployment, and finally they are released
        back to the pool or retired altogether.
      </p>
    </Strip>
  ),

  name: "Bordered strip",
};

export const DeepStrip: Story = {
  render: () => (
    <Strip
      deep
      includeCol={false}
      element="section"
      type="light"
      rowClassName="u-vertically-center"
    >
      <Col size={8}>
        <h2>The fastest way to go from development to production in IoT</h2>
        <p>
          Learn about how Ubuntu Core and snaps can help you build your
          connected devices.
        </p>
      </Col>
      <Col size={4} className="u-hide--small u-align--center">
        <img
          src="https://assets.ubuntu.com/v1/808a4e5b-iot.png?h=300"
          alt="Placeholder"
        />
      </Col>
    </Strip>
  ),

  name: "Deep strip",
};

export const ShallowStrip: Story = {
  render: () => (
    <Strip
      shallow
      includeCol={false}
      element="section"
      type="light"
      rowClassName="u-vertically-center"
    >
      <Col size={8}>
        <h2>The fastest way to go from development to production in IoT</h2>
        <p>
          Learn about how Ubuntu Core and snaps can help you build your
          connected devices.
        </p>
      </Col>
      <Col size={4} className="u-hide--small u-align--center">
        <img
          src="https://assets.ubuntu.com/v1/808a4e5b-iot.png?h=300"
          alt="Placeholder"
        />
      </Col>
    </Strip>
  ),

  name: "Shallow strip",
};
