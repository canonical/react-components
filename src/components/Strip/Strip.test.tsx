import { render, screen } from "@testing-library/react";
import React from "react";

import Strip from "./Strip";

describe("Strip ", () => {
  it("renders", () => {
    render(<Strip data-testid="strip">Test content</Strip>);
    expect(screen.getByTestId("strip")).toMatchSnapshot();
  });

  it("can set a background", () => {
    render(
      <Strip data-testid="strip" background="test.png">
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveStyle({
      backgroundImage: "url('test.png')",
    });
  });

  it("can set a type", () => {
    render(
      <Strip data-testid="strip" type="dark">
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveClass("p-strip--dark");
  });

  it("can set the element", () => {
    render(
      <Strip data-testid="strip" element="section">
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip").nodeName).toEqual("SECTION");
  });

  it("can be bordered", () => {
    render(
      <Strip data-testid="strip" bordered={true}>
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveClass("is-bordered");
  });

  it("can be dark", () => {
    render(
      <Strip data-testid="strip" dark={true}>
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveClass("is-dark");
  });

  it("can be deep", () => {
    render(
      <Strip data-testid="strip" deep={true}>
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveClass("is-deep");
  });

  it("can be light", () => {
    render(
      <Strip data-testid="strip" light={true}>
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveClass("is-light");
  });

  it("can be shallow", () => {
    render(
      <Strip data-testid="strip" shallow={true}>
        Test content
      </Strip>
    );
    expect(screen.getByTestId("strip")).toHaveClass("is-shallow");
  });

  it("can set a col size", () => {
    render(
      <Strip data-testid="strip" colSize={4}>
        Test content
      </Strip>
    );
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".row .col-4")).toBeInTheDocument();
  });

  it("can not include a col", () => {
    render(
      <Strip data-testid="strip" includeCol={false}>
        Test content
      </Strip>
    );
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".row .col-12")).not.toBeInTheDocument();
  });

  it("can set row classes", () => {
    render(
      <Strip data-testid="strip" rowClassName="row--extra">
        Test content
      </Strip>
    );
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".row")).toHaveClass("row--extra");
  });

  it("can add additional classes", () => {
    render(
      <Strip data-testid="strip" className="extra-class">
        Test content
      </Strip>
    );
    const strip = screen.getByTestId("strip");
    expect(strip).toHaveClass("p-strip");
    expect(strip).toHaveClass("extra-class");
  });
});
