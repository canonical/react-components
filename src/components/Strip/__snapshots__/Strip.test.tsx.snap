// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Strip  renders 1`] = `
<div
  class="p-strip"
  data-testid="strip"
>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      Test content
    </div>
  </div>
</div>
`;
